server/src/main/java/org/elasticsearch/search/vectors/KnnRescoreVectorQuery.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.queries.function.FunctionScoreQuery;
+import org.apache.lucene.search.DoubleValuesSource;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.elasticsearch.index.mapper.vectors.VectorSimilarityByteValueSource;
+import org.elasticsearch.index.mapper.vectors.VectorSimilarityFloatValueSource;
+import org.elasticsearch.search.profile.query.QueryProfiler;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Wraps a kNN vector query to rescore the results using the non-quantized vectors
+ */
+public class KnnRescoreVectorQuery extends Query implements ProfilingQuery {
+    private final String fieldName;
+    private final byte[] byteTarget;
+    private final float[] floatTarget;
+    private final VectorSimilarityFunction vectorSimilarityFunction;
+    private final Integer k;
+    private final Query vectorQuery;
+
+    private long vectorOpsCount;
+
+    public KnnRescoreVectorQuery(
+        String fieldName,
+        byte[] byteTarget,
+        VectorSimilarityFunction vectorSimilarityFunction,
+        Integer k,
+        Query vectorQuery
```

Comment:
the knn query should also be a profiled query

Reply:
Mmmm, gotcha. There can be multiple levels of wrapping:
- KnnRescoreVectorQuery
- VectorSimilarityQuery
- EsKnnFloatVectorQuery

We'll probably need `VectorSimilarityQuery` to implement `ProfilingQuery` as well - right now using a `similarity` param means that profiling returns 0 vector op counts.

I'll fix that, thanks for the catch 👍